### PR TITLE
all: use T.TempDir to create temporary test directories

### PR DESCRIPTION
**Pull request by Juneezee**
A testing cleanup. 

This pull request replaces `ioutil.TempDir` with `t.TempDir`. We can use the `T.TempDir` function from the `testing` package to create temporary directory. The directory created by `T.TempDir` is automatically removed when the test and all its subtests complete. 

Reference: https://pkg.go.dev/testing#T.TempDir

### CODE REVIEW

File: accounts/keystore/keystore_test.go [hunk 1]
```diff
@@ -462,10 +448,7 @@ func checkEvents(t *testing.T, want []walletEvent, have []walletEvent) {
 }
 
 func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
```
Comment (s1na) [inline accounts/keystore/keystore_test.go:450]: We don't need to return the dir here anymore

File: consensus/ethash/algorithm_test.go [hunk 2]
```diff
@@ -698,6 +698,8 @@ func TestHashimoto(t *testing.T) {
 // Tests that caches generated on disk may be done concurrently.
 func TestConcurrentDiskCacheGeneration(t *testing.T) {
 	// Create a temp folder to generate the caches into
+	// TODO: t.TempDir fails to remove the directory on Windows
```
Comment (s1na) [inline consensus/ethash/algorithm_test.go:701]: Do you know why is this a problem here but not in the other tests?

**Review by s1na — CHANGES_REQUESTED** [review]
Nice cleanup thanks. Also checked that `T.TempDir` still does cleanup in case of panic (https://github.com/golang/go/issues/41355). The PR also fixes some places where closing the db was missing. Just a few minor comments.

File: consensus/ethash/algorithm_test.go (hunk 2 above)
Comment (Juneezee) [reply consensus/ethash/algorithm_test.go:701]: This happens in `algorithm_test.go: TestConcurrentDiskCacheGeneration` and `ethash_test.go: TestCacheFileEvict` only, where we pass the temporary directory to `Config { CacheDir: tmpdir }`. 

Unfortunately, I don't have a Windows machine so I am unable to further debug this. It _might_ be a Windows specific issue, see https://github.com/golang/go/issues/50051

File: accounts/keystore/keystore_test.go [hunk 3]
```diff
@@ -38,7 +37,6 @@ var testSigData = make([]byte, 32)
 
 func TestKeyStore(t *testing.T) {
 	dir, ks := tmpKeyStore(t, true)
```
Comment (Juneezee) [inline accounts/keystore/keystore_test.go:39]: @s1na Reply for https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841913624. The `dir` is used here

File: accounts/keystore/account_cache_test.go [hunk 4]
```diff
@@ -55,7 +55,6 @@ func TestWatchNewFile(t *testing.T) {
 	t.Parallel()
 
 	dir, ks := tmpKeyStore(t, false)
```
Comment (Juneezee) [inline accounts/keystore/account_cache_test.go:57]: @s1na Reply for https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841913624. The `dir` is used here

File: accounts/keystore/keystore_test.go (hunk 1 above)
Comment (Juneezee) [reply accounts/keystore/keystore_test.go:450]: We still need to return the `dir` here. It is being used in two tests: https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841953459 and https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841953523

File: core/rawdb/freezer_test.go [hunk 5]
```diff
@@ -97,7 +95,6 @@ func TestFreezerModifyRollback(t *testing.T) {
 	t.Parallel()
 
 	f, dir := newFreezerForTesting(t, freezerTestTableDef)
```
Comment (Juneezee) [inline core/rawdb/freezer_test.go:97]: @s1na Reply for https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841935416. The `dir` is used here

File: core/rawdb/freezer_test.go [hunk 6]
```diff
@@ -295,10 +286,7 @@ func TestFreezerReadonlyValidate(t *testing.T) {
 func newFreezerForTesting(t *testing.T, tables map[string]bool) (*freezer, string) {
```
[the remark this answers is not in the document]
Comment (Juneezee) [reply core/rawdb/freezer_test.go:286]: The `dir` is being used in `TestFreezerModifyRollback`, so we still need to return the `dir` here. See https://github.com/ethereum/go-ethereum/pull/24633#discussion_r841957463

File: cmd/geth/run_test.go [hunk 7]
```diff
@@ -82,15 +73,10 @@ func runGeth(t *testing.T, args ...string) *testgeth {
 		}
 	}
 	if tt.Datadir == "" {
-		tt.Datadir = tmpdir(t)
-		tt.Cleanup = func() { os.RemoveAll(tt.Datadir) }
+		// The temporary datadir will be removed automatically if something fails below.
+		tt.Datadir = t.TempDir()
+		tt.Cleanup = func() {}
```
[the remark this answers is not in the document]
Comment (Juneezee) [reply cmd/geth/run_test.go]: Yup, we don't need it anymore. Removed in e7d4e74cdb6fad2cbbe05c4cf90c25e024c9dd30

**Review by s1na — APPROVED** [review]
Looks good!

**Review by fjl — APPROVED** [review]

**Pull request merged**